File: app/src/main/java/maderski/bluetoothautoplaymusic/Helpers/A2DPHelper.java
package maderski.bluetoothautoplaymusic.Helpers;

import android.bluetooth.BluetoothAdapter;
import android.bluetooth.BluetoothDevice;
import android.bluetooth.BluetoothProfile;
import android.content.Context;
import android.support.v4.util.ArraySet;
import android.util.Log;

import java.util.List;
import java.util.Set;

import maderski.bluetoothautoplaymusic.Utils.BluetoothDeviceUtils;

/**
 * Created by Jason on 6/25/17.
 */

public class A2DPHelper implements BluetoothProfile.ServiceListener{
    private static final String TAG = "A2DPHelper";

    public interface A2DPCallbacks {
        void connectedDeviceNames(Set<String> deviceNames);
    }

    private A2DPCallbacks mCallbacks;

    public A2DPHelper(A2DPCallbacks callbacks){
        mCallbacks = callbacks;
    }

    // Get List of currently connected bluetooth devices
    public void getConnectedA2DPDevices(Context context){
        BluetoothAdapter bluetoothAdapter = BluetoothAdapter.getDefaultAdapter();
        bluetoothAdapter.getProfileProxy(context, this, BluetoothProfile.A2DP);
    }

    @Override
    public void onServiceConnected(int profile, BluetoothProfile proxy) {
        Set<String> deviceNames = new ArraySet<>();
        List<BluetoothDevice> devices = proxy.getConnectedDevices();
        for(BluetoothDevice device : devices){
            Log.d(TAG, "CONNECTED DEVICE: " + device.getName());
            deviceNames.add(device.getName());
        }
        mCallbacks.connectedDeviceNames(deviceNames);
    }

    @Override
    public void onServiceDisconnected(int profile) {

    }
}

File: app/src/main/java/maderski/bluetoothautoplaymusic/helpers/A2DPHelper.kt
package maderski.bluetoothautoplaymusic.helpers

import android.bluetooth.BluetoothAdapter
import android.bluetooth.BluetoothProfile
import android.content.Context
import android.util.Log
import maderski.bluetoothautoplaymusic.bluetooth.models.BTDevice

/**
 * Created by Jason on 6/25/17.
 */

class A2DPHelper(private val callback: A2DPCallback) : BluetoothProfile.ServiceListener {

    interface A2DPCallback {
        fun connectedDeviceNames(deviceNames: Set<BTDevice>)
    }

    // Get List of currently connected bluetooth devices
    fun getConnectedA2DPDevices(context: Context) {
        val bluetoothAdapter = BluetoothAdapter.getDefaultAdapter()
        bluetoothAdapter.getProfileProxy(context, this, BluetoothProfile.A2DP)
    }

    override fun onServiceConnected(profile: Int, proxy: BluetoothProfile) {
        val deviceNames = proxy.connectedDevices.map{ device ->
            BTDevice(device.name, device.address)
        }.toSet()
        Log.d(TAG, "CONNECTED DEVICES: $deviceNames")
        callback.connectedDeviceNames(deviceNames)
    }

    override fun onServiceDisconnected(profile: Int) {}

    companion object {
        private const val TAG = "A2DPHelper"
    }
}
